src/Sentry/Internal/PollingNetworkStatusListener.cs

```diff
@@ -0,0 +1,70 @@
+using Sentry.Extensibility;
+
+namespace Sentry.Internal;
+
+internal class PollingNetworkStatusListener : INetworkStatusListener
+{
+    private readonly SentryOptions? _options;
+    private readonly IPing? _testPing;
+    internal int _delayInMilliseconds;
+    private readonly int _maxDelayInMilliseconds;
+    private readonly Func<int, int> _backoffFunction;
+
+    public PollingNetworkStatusListener(SentryOptions options, int initialDelayInMilliseconds = 500,
+        int maxDelayInMilliseconds = 32_000, Func<int, int>? backoffFunction = null)
+    {
+        _options = options;
+        _delayInMilliseconds = initialDelayInMilliseconds;
+        _maxDelayInMilliseconds = maxDelayInMilliseconds;
+        _backoffFunction = backoffFunction ?? (x => x * 2);
+    }
+
+    /// <summary>
+    /// Overload for testing
+    /// </summary>
+    internal PollingNetworkStatusListener(IPing testPing, int initialDelayInMilliseconds = 500,
+        int maxDelayInMilliseconds = 32_000, Func<int, int>? backoffFunction = null)
+    {
+        _testPing = testPing;
+        _delayInMilliseconds = initialDelayInMilliseconds;
+        _maxDelayInMilliseconds = maxDelayInMilliseconds;
+        _backoffFunction = backoffFunction ?? (x => x * 2);
+    }
+
+    private Lazy<IPing> LazyPing => new(() =>
+    {
+        if (_testPing != null)
+        {
+            return _testPing;
+        }
+        // If not running unit tests then _options will not be null and SDK init would fail without a Dsn
+        var uri = new Uri(_options!.Dsn!);
+        return new TcpPing(uri.DnsSafeHost, uri.Port);
+    });
+    private IPing Ping => LazyPing.Value;
+
+    private volatile bool _online = true;
+    public bool Online
+    {
+        get => _online;
+        set => _online = value;
+    }
+
+    public async Task WaitForNetworkOnlineAsync(CancellationToken cancellationToken = default)
```

Comment:
This argument has a default value but I don't see it being tested when a cancellation token is passed.

If one is passed, and triggered, `Delay` would throw an exception that isnt' handled anywhere here. Is that considered?

I'd recommend remove the argument if we're not using anywhere (at least I can't see it being used).

If we're using it, add a test that does that to verify the behavior

Reply:
Good call. Added test and modified code to ensure a graceful exit if the task is cancelled:
- https://github.com/getsentry/sentry-dotnet/pull/3506